### PR TITLE
Convenience Actions methods for Scrolling

### PR DESCRIPTION
**Pull request by titusfortner**
Trunk right now has basic scrolling support merged. 
This PR is set to conveniently show the diff of what exists and what could exist.

### Options for reviewers:

1. Reject all changes and keep what is in trunk exactly as it is
2. Reject PR but make changes to what is in trunk 
3. Accept PR, but make proposed changes
4. Accept PR and/or do nothing and this will get merged before 4.2 is released

### Currently in Trunk

* I updated the Ruby implementation match the Python implementation approach.
* I made the Java & .NET approaches similar (except .NET is using a class with properties, Java is using a class with static methods)
* All languages have the same descriptions for each of the parameters
* We have tests for the 5 working and 2 failing tests for these 4 languages.

There are legit issues to be taken with the implementation in trunk (especially with .NET since the class with properties isn't ideal). But, it is basic implementation of a `scroll()` method with all the functionality.

### Changes proposed by this PR

The primary reason for this PR is that I believe it is not intuitive for users to understand the purpose of `x` and `y`, even if we called them `offsetX` and `offsetY`. The idea of offset from origin and the origin being one of two different things took me a while to figure out, and I like to think I'm at least average in figuring out this stuff compared to our users.

This PR removes a generic scroll method and replaces it with 3 convenience methods; the first 2 represent what 98% of use cases require. The 3rd is the featured method but it makes it blatantly obvious that the offsets are explicitly associated with the origin (viewport or element).

The names match the existing language conventions for pointer:

1. Scroll to Element — takes element argument
.NET — `ScrollToElement()`
Java — `scrollToElement()`
Python — `scroll_to_element()`
Ruby — `#scroll_to`

2. Scroll by Amount — takes deltaX/deltaY values:
.NET — `ScrollByAmount()`
Java — `scrollByAmount()`
Python — `scroll_by_amount()`
Ruby — `#scroll_by`

3. Scroll from an origin — takes deltaX/deltaY and a ScrollOrigin object instance
.NET — `ScrollFromOrigin()`
Java — `scrollFromOrigin()`
Python — `scroll_from_origin()`
Ruby — `#scroll_from`

There are a lot of potential (and valid) quibbles with this implementation, from method names to defaults, to where defaults live, to which stage of the actions class process does translations of scroll origin, etc. I tried very hard to make the implementations happen at the same "place" in the different languages, but in a language idiomatic way.

### CODE REVIEW

File: py/selenium/webdriver/common/action_chains.py
Comment (symonk) [inline py/selenium/webdriver/common/action_chains.py:328]: I think we need to stagger this change over two releases? (its a breaking public API change I think), should we keep `scroll(...)` with a `DeprecationWarning`?

File: py/selenium/webdriver/common/actions/wheel_actions.py
Comment (symonk) [inline py/selenium/webdriver/common/actions/wheel_actions.py:32]: be nice to type these; `int/float?` `float` if either is acceptible

File: py/selenium/webdriver/common/actions/wheel_input.py
Comment (symonk) [inline py/selenium/webdriver/common/actions/wheel_input.py:32]: does this offer any value over the normal init? (can the dunder `__init__` have defaults for the offsets as `=0` itself?

File: py/selenium/webdriver/common/actions/wheel_input.py
Comment (symonk) [inline py/selenium/webdriver/common/actions/wheel_input.py:40]: read-only ok for these? (seems ok to me)

File: py/selenium/webdriver/common/action_chains.py
Comment (AutomatedTester) [reply py/selenium/webdriver/common/action_chains.py:328]: We can't break APIs like this.

Comment (titusfortner) [reply py/selenium/webdriver/common/action_chains.py:328]: This is the downside of doing patch releases instead of waiting for minor releases with new code, but yes, can keep existing method and mark it deprecated instead of deleting it.

File: py/selenium/webdriver/common/actions/wheel_actions.py
Comment (titusfortner) [reply py/selenium/webdriver/common/actions/wheel_actions.py:32]: I had types and then removed them. I want to allow comparing apples to apples for this PR. We're going to do a full update on typing in the future.

File: py/selenium/webdriver/common/actions/wheel_input.py
Comment (titusfortner) [reply py/selenium/webdriver/common/actions/wheel_input.py:32]: The idea is to make it explicit the object you get instead of just constructing by default. This is a common Java pattern that I also use in Ruby. If it isn't idiomatic Python we can adjust. Java can make the constructor private. Ruby & Python people can avoid the class methods, but we can at least have them for our documentation.

File: py/selenium/webdriver/common/actions/wheel_input.py
Comment (titusfortner) [reply py/selenium/webdriver/common/actions/wheel_input.py:40]: Yes, I thought you had to explicitly add a setter to make it not read only?

File: py/selenium/webdriver/common/action_chains.py
Comment (AutomatedTester) [reply py/selenium/webdriver/common/action_chains.py:328]: We can't break APIs like this.

**Review by AutomatedTester — CHANGES_REQUESTED** [review]
APIs can't just change from under people in any of the code changes. That goes against the usual deprecation policies we have.

File: py/selenium/webdriver/common/actions/wheel_input.py
Comment (symonk) [reply py/selenium/webdriver/common/actions/wheel_input.py:40]: Yep I'm just wondering if that was the intent (vs allowing them to be set and modified afterwards on the instance)

**Comment by titusfortner** [comment]
Yes, I guess I kind of would like to treat changes made in patches as different from changes made (and announced) in minor versions. Otherwise we need to rethink feature branch management. Or just give up on trying to collaborate between bindings.

The fix in this case is easy enough, though. I restored the previous python method and marked it as deprecated.
None of the other languages have had a patch release since the changed method/class names were introduced.

File: py/selenium/webdriver/common/actions/wheel_input.py
Comment (titusfortner) [reply py/selenium/webdriver/common/actions/wheel_input.py:40]: Ideally the constructor would also be "private" like we can do it in Java. Ruby we are marking the constructor `@api private` to annotate that. Not sure if there is an equivalent in Python. The ideal is to only create the instance via class method, which properly sets appropriate values/defaults.

**Comment by titusfortner** [comment]
Thinking more about the bigger picture here. I definitely overestimated what we'd be able to do back in November, which we need to better account for going forward.

TL/DR: I'd like to see us better prioritize 3 things:
1. Smaller more frequent minor releases (easier said than done)
2. Marketing pushes w/ blog posts for each minor release (requires languages to each have the feature implemented)
3. More consistency between the OOP languages (JS is a little special; the rest can do more similar approaches even in language idiomatic ways)

Selenium needs better marketing. That means making blog posts publicizing the cool new features in minor releases and pointing to good, working examples for each of them and encouraging people to upgrade. We need people to be talking and blogging and speaking about Selenium by constantly keeping ourselves worth talking about.

For that to work well, we kind of need the bindings in lock-step. I know this was never a priority in 3.x and each language kind of released features whenever they realized they needed to be added and the amount of communication wasn't ideal. I'd like to prioritize not doing this.

But it makes things weird when it comes to things like:
https://github.com/SeleniumHQ/selenium/issues/10541
Do we need to get the other languages to implement it before we can release 4.2 if we want to advertise it? Do we need to collaborate better on new features and not merge things until everyone is ready to merge and release a new minor version?
Do we maybe do patch releases as beta releases so we can change the API before a minor release while waiting for everyone to implement? (I'd rather figure out how to do smaller more frequent (monthly w/Chrome updates?) minor releases, but that's hard with current resources. 

As for consistency, I'm coming at this from the standpoint of trying to write examples for Sauce Labs and then for our Selenium documentation. It would be much easier to show smaller code snippets if the general approach were kept a little more similar. im did a really good job of keeping .NET functionality similar to Java, though Java didn't always keep up with .NET. I definitely think the Ruby bindings are lacking because when we (I) implemented things, I didn't take any cues from Java or .NET in a way that would have been helpful looking at it in retrospect.

Anyway, my Sunday morning thoughts.

**Comment by titusfortner** [comment]
Here's the publicly executable test code and the documentation referencing it, if we merge this PR:
https://deploy-preview-1010--jovial-austin-42fe02.netlify.app/documentation/webdriver/actions_api/wheel/

File: py/selenium/webdriver/common/action_chains.py
Comment (diemol) [reply py/selenium/webdriver/common/action_chains.py:328]: Seems you a had an issue in your comments @AutomatedTester, somehow it went twice :)

Comment (AutomatedTester) [reply py/selenium/webdriver/common/action_chains.py:328]: Weird... It wasn't intended to be twice

Comment (titusfortner) [reply py/selenium/webdriver/common/action_chains.py:328]: I think we posted at the same time and github got confused how to display things.

**Comment by pujagani** [comment]
Thank you, Titus! Java code changes look good to me. Can you please help me understand why the Ruby method names differ from other bindings? Apologies, I might be missing some context here.

**Comment by titusfortner** [comment]
Great question. Each language is matching the previous naming conventions from mouse movement. All the other languages have longer methods names, ruby truncated the method name and relies on the first parameter to complete the idea.

**Pull request merged**
